File: worker/app/license.py
import base64
import hashlib
import json
import os
import threading
from datetime import datetime, timezone
from typing import Any, Dict, Optional

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding

from app import db
from app.local_testing_state import get_local_testing_state
from app.runtime_config import is_local_docker_deployment


LICENSE_SCHEMA_VERSION = 1
LICENSE_SIGNATURE_DELIMITER = "\n---SIGNATURE---\n"
LICENSE_FEATURE_DEFAULTS = {
    "dashboard_read": True,
    "live_graph_read": True,
    "admin_view": True,
    "license_manage": True,
    "permission_revoke": False,
    "job_control": False,
    "graph_ingest": False,
    "copilot_telemetry": False,
    "agents_dashboard": False,
}
JOB_TYPE_LICENSE_FEATURES = {
    "graph_ingest": "graph_ingest",
    "copilot_telemetry": "copilot_telemetry",
}

_cache_lock = threading.Lock()
_summary_cache = None
_public_key_cache = None


class LicenseFeatureError(RuntimeError):
    def __init__(self, feature_key: str, summary: Dict[str, Any]):
        super().__init__(f"license_feature_{feature_key}_disabled")
        self.feature_key = feature_key
        self.summary = summary


def _normalize_newlines(value: str) -> str:
    return value.replace("\r\n", "\n")


def _normalize_timestamp(value: Optional[Any]) -> Optional[str]:
    if value is None:
        return None
    if isinstance(value, datetime):
        return _format_canonical_timestamp(value)
    parsed = _parse_datetime(str(value))
    return _format_canonical_timestamp(parsed) if parsed else str(value)


def _sort_json_value(value: Any) -> Any:
    if isinstance(value, list):
        return [_sort_json_value(entry) for entry in value]
    if isinstance(value, dict):
        return {key: _sort_json_value(value[key]) for key in sorted(value.keys())}
    return value


def canonicalize_license_payload(value: Any) -> str:
    return json.dumps(_sort_json_value(value), separators=(",", ":"), ensure_ascii=False)


def _default_features() -> Dict[str, bool]:
    return dict(LICENSE_FEATURE_DEFAULTS)


def _fully_enabled_features() -> Dict[str, bool]:
    return {key: True for key in LICENSE_FEATURE_DEFAULTS.keys()}


def _fallback_summary(error: Optional[str], **overrides) -> Dict[str, Any]:
    summary = {
        "status": "invalid",
        "mode": "read_only",
        "verification_status": "invalid",
        "verification_error": error,
        "artifact_id": None,
        "sha256": None,
        "uploaded_at": None,
        "uploaded_by": {"oid": None, "upn": None, "name": None},
        "payload": None,
        "features": _default_features(),
    }
    summary.update(overrides)
    return summary


def _local_docker_license_summary() -> Dict[str, Any]:
    tenant_id = _non_empty_string(os.getenv("ENTRA_TENANT_ID")) or "local-docker"
    features = _fully_enabled_features()
    return {
        "status": "active",
        "mode": "full",
        "verification_status": "verified",
        "verification_error": None,
        "artifact_id": None,
        "sha256": None,
        "uploaded_at": None,
        "uploaded_by": {"oid": None, "upn": None, "name": None},
        "payload": {
            "schema_version": LICENSE_SCHEMA_VERSION,
            "license_id": "local-docker-emulated-license",
            "license_type": "local_docker",
            "tenant_id": tenant_id,
            "issued_at": "1970-01-01T00:00:00.000Z",
            "expires_at": None,
            "features": features,
        },
        "features": features,
    }


def get_license_lookup_failure_summary(error: Optional[str] = None) -> Dict[str, Any]:
    return _fallback_summary(error or "license_lookup_failed")


def _non_empty_string(value: Any) -> Optional[str]:
    if not isinstance(value, str):
        return None
    normalized = value.strip()
    return normalized or None


def _parse_datetime(value: Optional[str]) -> Optional[datetime]:
    if value is None:
        return None
    normalized = value.strip()
    if not normalized:
        return None
    try:
        return datetime.fromisoformat(normalized.replace("Z", "+00:00")).astimezone(timezone.utc)
    except ValueError:
        return None


def _format_canonical_timestamp(value: datetime) -> str:
    return value.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _parse_required_timestamp(value: Any, field_name: str, allow_null: bool = False) -> Optional[str]:
    if value is None and allow_null:
        return None
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{field_name}_required")
    parsed = _parse_datetime(value)
    if parsed is None:
        raise ValueError(f"{field_name}_invalid")
    return _format_canonical_timestamp(parsed)


def _derive_features(value: Any) -> Dict[str, bool]:
    if not isinstance(value, dict):
        raise ValueError("license_features_invalid")
    features = _default_features()
    for key in LICENSE_FEATURE_DEFAULTS.keys():
        if key not in value:
            continue
        raw = value[key]
        if not isinstance(raw, bool):
            raise ValueError(f"license_feature_{key}_invalid")
        features[key] = raw
    return features


def _parse_payload(value: Any) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError("license_payload_invalid")
    if value.get("schema_version") != LICENSE_SCHEMA_VERSION:
        raise ValueError("license_schema_version_invalid")
    license_id = _non_empty_string(value.get("license_id"))
    license_type = _non_empty_string(value.get("license_type"))
    tenant_id = _non_empty_string(value.get("tenant_id"))
    if not license_id:
        raise ValueError("license_id_required")
    if not license_type:
        raise ValueError("license_type_required")
    if not tenant_id:
        raise ValueError("license_tenant_id_required")
    return {
        "schema_version": LICENSE_SCHEMA_VERSION,
        "license_id": license_id,
        "license_type": license_type,
        "tenant_id": tenant_id,
        "issued_at": _parse_required_timestamp(value.get("issued_at"), "license_issued_at"),
        "expires_at": _parse_required_timestamp(value.get("expires_at"), "license_expires_at", allow_null=True),
        "features": _derive_features(value.get("features")),
    }


def _load_public_key():
    global _public_key_cache
    public_key_path = _non_empty_string(os.getenv("LICENSE_PUBLIC_KEY_PATH"))
    if not public_key_path:
        raise RuntimeError("license_public_key_path_not_configured")
    with _cache_lock:
        if _public_key_cache and _public_key_cache["path"] == public_key_path:
            return _public_key_cache["value"]
        with open(public_key_path, "rb") as handle:
            value = serialization.load_pem_public_key(handle.read())
        _public_key_cache = {"path": public_key_path, "value": value}
        return value


def inspect_license_artifact(raw_license_text: str) -> Dict[str, Any]:
    sha256 = hashlib.sha256(raw_license_text.encode("utf-8")).hexdigest()
    normalized = _normalize_newlines(raw_license_text)
    try:
        delimiter_index = normalized.find(LICENSE_SIGNATURE_DELIMITER)
        if delimiter_index < 0:
            raise ValueError("license_signature_delimiter_missing")
        if normalized.find(LICENSE_SIGNATURE_DELIMITER, delimiter_index + len(LICENSE_SIGNATURE_DELIMITER)) >= 0:
            raise ValueError("license_signature_delimiter_duplicated")

        payload_text = normalized[:delimiter_index]
        signature_text = normalized[delimiter_index + len(LICENSE_SIGNATURE_DELIMITER) :].strip()
        if not signature_text:
          raise ValueError("license_signature_missing")
        if any(ch.isspace() for ch in signature_text):
            raise ValueError("license_signature_invalid")

        payload = _parse_payload(json.loads(payload_text))
        canonical_payload = canonicalize_license_payload(payload)
        if payload_text != canonical_payload:
            raise ValueError("license_payload_not_canonical")

        public_key = _load_public_key()
        signature = base64.b64decode(signature_text.encode("ascii"), validate=True)
        public_key.verify(signature, canonical_payload.encode("utf-8"), padding.PKCS1v15(), hashes.SHA256())
        return {
            "verification_status": "verified",
            "verification_error": None,
            "payload": payload,
            "sha256": sha256,
        }
    except Exception as exc:
        return {
            "verification_status": "invalid",
            "verification_error": str(exc) or "license_verification_failed",
            "payload": None,
            "sha256": sha256,
        }


def summarize_license_artifact(raw_license_text: str, *, artifact_id=None, uploaded_at=None, uploaded_by=None) -> Dict[str, Any]:
    inspection = inspect_license_artifact(raw_license_text)
    metadata = {
        "artifact_id": artifact_id,
        "sha256": inspection["sha256"],
        "uploaded_at": _normalize_timestamp(uploaded_at),
        "uploaded_by": {
            "oid": (uploaded_by or {}).get("oid"),
            "upn": (uploaded_by or {}).get("upn"),
            "name": (uploaded_by or {}).get("name"),
        },
    }

    if inspection["verification_status"] != "verified" or not inspection["payload"]:
        return _fallback_summary(
            inspection["verification_error"],
            artifact_id=metadata["artifact_id"],
            sha256=metadata["sha256"],
            uploaded_at=metadata["uploaded_at"],
            uploaded_by=metadata["uploaded_by"],
        )

    tenant_id = _non_empty_string(os.getenv("ENTRA_TENANT_ID"))
    if not tenant_id:
        return _fallback_summary(
            "entra_tenant_id_not_configured",
            artifact_id=metadata["artifact_id"],
            sha256=metadata["sha256"],
            uploaded_at=metadata["uploaded_at"],
            uploaded_by=metadata["uploaded_by"],
            payload=inspection["payload"],
        )

    if inspection["payload"]["tenant_id"] != tenant_id:
        return _fallback_summary(
            "license_tenant_id_mismatch",
            artifact_id=metadata["artifact_id"],
            sha256=metadata["sha256"],
            uploaded_at=metadata["uploaded_at"],
            uploaded_by=metadata["uploaded_by"],
            payload=inspection["payload"],
        )

    expires_at = inspection["payload"]["expires_at"]
    expires_dt = _parse_datetime(expires_at) if expires_at else None
    if expires_dt is not None and expires_dt < datetime.now(timezone.utc):
        return {
            "status": "expired",
            "mode": "read_only",
            "verification_status": "verified",
            "verification_error": "license_expired",
            "artifact_id": metadata["artifact_id"],
            "sha256": metadata["sha256"],
            "uploaded_at": metadata["uploaded_at"],
            "uploaded_by": metadata["uploaded_by"],
            "payload": inspection["payload"],
            "features": _default_features(),
        }

    return {
        "status": "active",
        "mode": "full",
        "verification_status": "verified",
        "verification_error": None,
        "artifact_id": metadata["artifact_id"],
        "sha256": metadata["sha256"],
        "uploaded_at": metadata["uploaded_at"],
        "uploaded_by": metadata["uploaded_by"],
        "payload": inspection["payload"],
        "features": inspection["payload"]["features"],
    }


def clear_license_cache():
    global _summary_cache
    with _cache_lock:
        _summary_cache = None


def get_current_license() -> Dict[str, Any]:
    global _summary_cache
    ttl_seconds = int(os.getenv("LICENSE_CACHE_TTL_SECONDS", "300") or "300")
    local_testing_state = get_local_testing_state() if is_local_docker_deployment() else None
    meta = None
    if local_testing_state is None:
        meta = db.fetch_one(
            """
            SELECT ala.artifact_id, ala.updated_at, la.sha256
            FROM active_license_artifact ala
            LEFT JOIN license_artifacts la ON la.artifact_id = ala.artifact_id
            WHERE ala.slot = 'default'
            LIMIT 1
            """
        )
    cache_key = (
        "local:%s:%s"
        % (
            "enabled" if local_testing_state and local_testing_state["emulate_license_enabled"] else "disabled",
            local_testing_state.get("updated_at") or "none",
        )
        if local_testing_state is not None
        else "%s:%s:%s"
        % (
            meta.get("artifact_id") if meta else "missing",
            meta.get("sha256") if meta else "none",
            _normalize_timestamp(meta.get("updated_at")) if meta else "none",
        )
    )
    now_ts = datetime.now(timezone.utc).timestamp()

    with _cache_lock:
        if _summary_cache and _summary_cache["cache_key"] == cache_key and _summary_cache["expires_at_ts"] > now_ts:
            return _summary_cache["summary"]

    if local_testing_state is not None:
        summary = (
            _local_docker_license_summary()
            if local_testing_state["emulate_license_enabled"]
            else _fallback_summary("license_missing", status="missing", verification_status="missing")
        )
    elif not meta or not meta.get("artifact_id"):
        summary = _fallback_summary("license_missing", status="missing", verification_status="missing")
    else:
        artifact = db.fetch_one(
            """
            SELECT artifact_id,
                   raw_license_text,
                   sha256,
                   uploaded_by_oid,
                   uploaded_by_upn,
                   uploaded_by_name,
                   uploaded_at,
                   verification_status,
                   verification_error
            FROM license_artifacts
            WHERE artifact_id = %s
            LIMIT 1
            """,
            [meta["artifact_id"]],
        )
        if not artifact:
            summary = _fallback_summary(
                "active_license_artifact_missing_row",
                artifact_id=meta.get("artifact_id"),
                sha256=meta.get("sha256"),
            )
        else:
            summary = summarize_license_artifact(
                artifact["raw_license_text"],
                artifact_id=artifact["artifact_id"],
                uploaded_at=artifact.get("uploaded_at"),
                uploaded_by={
                    "oid": artifact.get("uploaded_by_oid"),
                    "upn": artifact.get("uploaded_by_upn"),
                    "name": artifact.get("uploaded_by_name"),
                },
            )

    with _cache_lock:
        _summary_cache = {
            "cache_key": cache_key,
            "expires_at_ts": now_ts + max(1, ttl_seconds),
            "summary": summary,
        }
    return summary


def require_license_feature(feature_key: str) -> Dict[str, Any]:
    summary = get_current_license()
    if summary["features"].get(feature_key):
        return summary
    raise LicenseFeatureError(feature_key, summary)


def get_job_type_license_feature(job_type: Optional[str]) -> Optional[str]:
    if not job_type:
        return None
    return JOB_TYPE_LICENSE_FEATURES.get(str(job_type))

File: web/app/lib/license.ts
import { createHash, createVerify } from "node:crypto";
import { readFile } from "node:fs/promises";
import type { PoolClient } from "pg";

import { getLocalTestingState } from "./local-testing-state";
import { isLocalDockerDeployment } from "./runtime";

export const LICENSE_SCHEMA_VERSION = 1;
export const LICENSE_SIGNATURE_DELIMITER = "\n---SIGNATURE---\n";

export const LICENSE_FEATURE_DEFAULTS = {
  dashboard_read: true,
  live_graph_read: true,
  admin_view: true,
  license_manage: true,
  permission_revoke: false,
  job_control: false,
  graph_ingest: false,
  copilot_telemetry: false,
  agents_dashboard: true,
  copilot_usage_sync: false,
  copilot_dashboard: true,
} as const;

export type LicenseFeatureKey = keyof typeof LICENSE_FEATURE_DEFAULTS;
export type LicenseFeatures = Record<LicenseFeatureKey, boolean>;

export type LicensePayload = {
  schema_version: number;
  license_id: string;
  license_type: string;
  tenant_id: string;
  issued_at: string;
  expires_at: string | null;
  features: LicenseFeatures;
};

export type LicenseStatus = "active" | "expired" | "invalid" | "missing";
export type LicenseMode = "full" | "read_only";
export type LicenseVerificationStatus = "verified" | "invalid" | "missing";

export type EffectiveLicenseSummary = {
  status: LicenseStatus;
  mode: LicenseMode;
  verificationStatus: LicenseVerificationStatus;
  verificationError: string | null;
  artifactId: string | null;
  sha256: string | null;
  uploadedAt: string | null;
  uploadedBy: {
    oid: string | null;
    upn: string | null;
    name: string | null;
  };
  payload: LicensePayload | null;
  features: LicenseFeatures;
};

type LicenseArtifactRow = {
  artifact_id: string;
  raw_license_text: string;
  sha256: string;
  uploaded_by_oid: string | null;
  uploaded_by_upn: string | null;
  uploaded_by_name: string | null;
  uploaded_at: string | Date | null;
  verification_status: string;
  verification_error: string | null;
};

type ActiveLicenseMetaRow = {
  artifact_id: string | null;
  updated_at: string | Date | null;
  sha256: string | null;
};

type UploadedArtifactActor = {
  oid?: string | null;
  upn?: string | null;
  name?: string | null;
};

type LicenseInspection = {
  verificationStatus: LicenseVerificationStatus;
  verificationError: string | null;
  payload: LicensePayload | null;
  sha256: string;
};

type SummaryMetadata = {
  artifactId: string | null;
  uploadedAt: string | Date | null;
  uploadedBy: {
    oid: string | null;
    upn: string | null;
    name: string | null;
  };
  sha256: string;
};

type SummaryOverride = EffectiveLicenseSummary | (() => Promise<EffectiveLicenseSummary>);
type LicenseQueryFn = <T = any>(text: string, params?: any[]) => Promise<T[]>;

export class LicenseFeatureError extends Error {
  featureKey: LicenseFeatureKey;
  summary: EffectiveLicenseSummary;

  constructor(featureKey: LicenseFeatureKey, summary: EffectiveLicenseSummary) {
    super(`license_feature_${featureKey}_disabled`);
    this.featureKey = featureKey;
    this.summary = summary;
  }
}

let summaryOverride: SummaryOverride | null = null;
let publicKeyOverride: string | null = null;
let licenseQueryOverride: LicenseQueryFn | null = null;
let cachedSummary:
  | {
      cacheKey: string;
      expiresAtMs: number;
      summary: EffectiveLicenseSummary;
    }
  | null = null;
let cachedPublicKey:
  | {
      path: string;
      value: string;
    }
  | null = null;

async function runLicenseQuery<T = any>(text: string, params?: any[]): Promise<T[]> {
  if (licenseQueryOverride) {
    return licenseQueryOverride<T>(text, params);
  }

  const db = await import("./db");
  return db.query<T>(text, params);
}

async function runLicenseTransaction<T>(fn: (client: PoolClient) => Promise<T>): Promise<T> {
  const db = await import("./db");
  return db.withTransaction(fn);
}

function normalizeNewlines(value: string): string {
  return value.replace(/\r\n/g, "\n");
}

function normalizeTimestamp(value: string | Date | null | undefined): string | null {
  if (!value) {
    return null;
  }
  if (value instanceof Date) {
    return value.toISOString();
  }
  const parsed = new Date(value);
  if (Number.isNaN(parsed.getTime())) {
    return String(value);
  }
  return parsed.toISOString();
}

function sortJsonValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((entry) => sortJsonValue(entry));
  }
  if (value && typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>).sort(([left], [right]) => left.localeCompare(right));
    return Object.fromEntries(entries.map(([key, entry]) => [key, sortJsonValue(entry)]));
  }
  return value;
}

export function canonicalizeLicensePayload(value: unknown): string {
  return JSON.stringify(sortJsonValue(value));
}

function getLicenseCacheTtlMs() {
  const raw = Number(process.env.LICENSE_CACHE_TTL_SECONDS || "300");
  if (!Number.isFinite(raw) || raw <= 0) {
    return 300_000;
  }
  return Math.floor(raw * 1000);
}

function getDefaultLicenseFeatures(): LicenseFeatures {
  return { ...LICENSE_FEATURE_DEFAULTS };
}

function getDisabledLicenseFeatures(): LicenseFeatures {
  return Object.fromEntries(
    (Object.keys(LICENSE_FEATURE_DEFAULTS) as LicenseFeatureKey[]).map((key) => [key, false])
  ) as LicenseFeatures;
}

function getFullyEnabledLicenseFeatures(): LicenseFeatures {
  return Object.fromEntries(
    (Object.keys(LICENSE_FEATURE_DEFAULTS) as LicenseFeatureKey[]).map((key) => [key, true])
  ) as LicenseFeatures;
}

function getReadOnlyFallbackSummary(error: string | null, overrides?: Partial<EffectiveLicenseSummary>): EffectiveLicenseSummary {
  return {
    status: overrides?.status ?? "invalid",
    mode: "read_only",
    verificationStatus: overrides?.verificationStatus ?? (overrides?.status === "missing" ? "missing" : "invalid"),
    verificationError: error,
    artifactId: overrides?.artifactId ?? null,
    sha256: overrides?.sha256 ?? null,
    uploadedAt: overrides?.uploadedAt ?? null,
    uploadedBy: overrides?.uploadedBy ?? { oid: null, upn: null, name: null },
    payload: overrides?.payload ?? null,
    features: overrides?.features ?? getDefaultLicenseFeatures(),
  };
}

function getLocalDockerLicenseSummary(): EffectiveLicenseSummary {
  const tenantId = asNonEmptyString(process.env.ENTRA_TENANT_ID) || "local-docker";
  const features = getFullyEnabledLicenseFeatures();

  return {
    status: "active",
    mode: "full",
    verificationStatus: "verified",
    verificationError: null,
    artifactId: null,
    sha256: null,
    uploadedAt: null,
    uploadedBy: { oid: null, upn: null, name: null },
    payload: {
      schema_version: LICENSE_SCHEMA_VERSION,
      license_id: "local-docker-emulated-license",
      license_type: "local_docker",
      tenant_id: tenantId,
      issued_at: "1970-01-01T00:00:00.000Z",
      expires_at: null,
      features,
    },
    features,
  };
}

function asNonEmptyString(value: unknown): string | null {
  if (typeof value !== "string") {
    return null;
  }
  const normalized = value.trim();
  return normalized ? normalized : null;
}

function parseIsoTimestamp(value: unknown, fieldName: string, allowNull = false): string | null {
  if (value === null && allowNull) {
    return null;
  }
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${fieldName}_required`);
  }
  const parsed = new Date(value);
  if (Number.isNaN(parsed.getTime())) {
    throw new Error(`${fieldName}_invalid`);
  }
  return parsed.toISOString();
}

function deriveLicenseFeatures(value: unknown): LicenseFeatures {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("license_features_invalid");
  }
  const merged = getDisabledLicenseFeatures();
  for (const key of Object.keys(LICENSE_FEATURE_DEFAULTS) as LicenseFeatureKey[]) {
    const raw = (value as Record<string, unknown>)[key];
    if (raw === undefined) {
      continue;
    }
    if (typeof raw !== "boolean") {
      throw new Error(`license_feature_${key}_invalid`);
    }
    merged[key] = raw;
  }
  return merged;
}

function parseLicensePayloadObject(value: unknown): LicensePayload {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("license_payload_invalid");
  }

  const payload = value as Record<string, unknown>;
  const schemaVersion = payload.schema_version;
  if (schemaVersion !== LICENSE_SCHEMA_VERSION) {
    throw new Error("license_schema_version_invalid");
  }

  const licenseId = asNonEmptyString(payload.license_id);
  if (!licenseId) {
    throw new Error("license_id_required");
  }
  const licenseType = asNonEmptyString(payload.license_type);
  if (!licenseType) {
    throw new Error("license_type_required");
  }
  const tenantId = asNonEmptyString(payload.tenant_id);
  if (!tenantId) {
    throw new Error("license_tenant_id_required");
  }

  return {
    schema_version: LICENSE_SCHEMA_VERSION,
    license_id: licenseId,
    license_type: licenseType,
    tenant_id: tenantId,
    issued_at: parseIsoTimestamp(payload.issued_at, "license_issued_at")!,
    expires_at: parseIsoTimestamp(payload.expires_at, "license_expires_at", true),
    features: deriveLicenseFeatures(payload.features),
  };
}

async function loadLicensePublicKey(): Promise<string> {
  if (publicKeyOverride !== null) {
    return publicKeyOverride;
  }

  const publicKeyPath = asNonEmptyString(process.env.LICENSE_PUBLIC_KEY_PATH);
  if (!publicKeyPath) {
    throw new Error("license_public_key_path_not_configured");
  }

  if (cachedPublicKey?.path === publicKeyPath) {
    return cachedPublicKey.value;
  }

  const publicKey = await readFile(publicKeyPath, "utf-8");
  cachedPublicKey = { path: publicKeyPath, value: publicKey };
  return publicKey;
}

export async function inspectLicenseArtifact(rawLicenseText: string): Promise<LicenseInspection> {
  const sha256 = createHash("sha256").update(rawLicenseText, "utf8").digest("hex");
  const normalized = normalizeNewlines(rawLicenseText);

  try {
    const delimiterIndex = normalized.indexOf(LICENSE_SIGNATURE_DELIMITER);
    if (delimiterIndex < 0) {
      throw new Error("license_signature_delimiter_missing");
    }
    if (normalized.indexOf(LICENSE_SIGNATURE_DELIMITER, delimiterIndex + LICENSE_SIGNATURE_DELIMITER.length) >= 0) {
      throw new Error("license_signature_delimiter_duplicated");
    }

    const payloadText = normalized.slice(0, delimiterIndex);
    const signatureText = normalized.slice(delimiterIndex + LICENSE_SIGNATURE_DELIMITER.length).trim();
    if (!signatureText) {
      throw new Error("license_signature_missing");
    }
    if (/\s/.test(signatureText)) {
      throw new Error("license_signature_invalid");
    }

    const rawPayload = JSON.parse(payloadText);
    const parsedPayload = parseLicensePayloadObject(rawPayload);
    const canonicalPayload = canonicalizeLicensePayload(rawPayload);
    if (payloadText !== canonicalPayload) {
      throw new Error("license_payload_not_canonical");
    }

    const publicKey = await loadLicensePublicKey();
    const verifier = createVerify("RSA-SHA256");
    verifier.update(canonicalPayload, "utf8");
    verifier.end();

    const validSignature = verifier.verify(publicKey, Buffer.from(signatureText, "base64"));
    if (!validSignature) {
      throw new Error("license_signature_invalid");
    }

    return {
      verificationStatus: "verified",
      verificationError: null,
      payload: parsedPayload,
      sha256,
    };
  } catch (error: unknown) {
    const message = error instanceof Error ? error.message : "license_verification_failed";
    return {
      verificationStatus: "invalid",
      verificationError: message,
      payload: null,
      sha256,
    };
  }
}

function summarizeInspection(inspection: LicenseInspection, metadata: SummaryMetadata): EffectiveLicenseSummary {
  if (inspection.verificationStatus !== "verified" || !inspection.payload) {
    return getReadOnlyFallbackSummary(inspection.verificationError, {
      artifactId: metadata.artifactId,
      sha256: metadata.sha256,
      uploadedAt: normalizeTimestamp(metadata.uploadedAt),
      uploadedBy: metadata.uploadedBy,
    });
  }

  const tenantId = asNonEmptyString(process.env.ENTRA_TENANT_ID);
  if (!tenantId) {
    return getReadOnlyFallbackSummary("entra_tenant_id_not_configured", {
      artifactId: metadata.artifactId,
      sha256: metadata.sha256,
      uploadedAt: normalizeTimestamp(metadata.uploadedAt),
      uploadedBy: metadata.uploadedBy,
      payload: inspection.payload,
    });
  }

  if (inspection.payload.tenant_id !== tenantId) {
    return getReadOnlyFallbackSummary("license_tenant_id_mismatch", {
      artifactId: metadata.artifactId,
      sha256: metadata.sha256,
      uploadedAt: normalizeTimestamp(metadata.uploadedAt),
      uploadedBy: metadata.uploadedBy,
      payload: inspection.payload,
    });
  }

  const expired = inspection.payload.expires_at !== null && new Date(inspection.payload.expires_at).getTime() < Date.now();

  return {
    status: expired ? "expired" : "active",
    mode: expired ? "read_only" : "full",
    verificationStatus: "verified",
    verificationError: expired ? "license_expired" : null,
    artifactId: metadata.artifactId,
    sha256: metadata.sha256,
    uploadedAt: normalizeTimestamp(metadata.uploadedAt),
    uploadedBy: metadata.uploadedBy,
    payload: inspection.payload,
    features: expired ? getDefaultLicenseFeatures() : inspection.payload.features,
  };
}

export async function summarizeLicenseArtifactText(params: {
  rawLicenseText: string;
  artifactId?: string | null;
  uploadedAt?: string | Date | null;
  uploadedBy?: {
    oid?: string | null;
    upn?: string | null;
    name?: string | null;
  } | null;
}): Promise<EffectiveLicenseSummary> {
  const inspection = await inspectLicenseArtifact(params.rawLicenseText);
  return summarizeInspection(inspection, {
    artifactId: params.artifactId ?? null,
    uploadedAt: params.uploadedAt ?? null,
    uploadedBy: {
      oid: params.uploadedBy?.oid ?? null,
      upn: params.uploadedBy?.upn ?? null,
      name: params.uploadedBy?.name ?? null,
    },
    sha256: inspection.sha256,
  });
}

function summarizeArtifactRow(row: LicenseArtifactRow): Promise<EffectiveLicenseSummary> {
  return inspectLicenseArtifact(row.raw_license_text).then((inspection) =>
    summarizeInspection(inspection, {
      artifactId: row.artifact_id,
      sha256: inspection.sha256,
      uploadedAt: row.uploaded_at,
      uploadedBy: {
        oid: row.uploaded_by_oid,
        upn: row.uploaded_by_upn,
        name: row.uploaded_by_name,
      },
    })
  );
}

async function getActiveLicenseMeta(): Promise<ActiveLicenseMetaRow | null> {
  const rows = await runLicenseQuery<ActiveLicenseMetaRow>(
    `
    SELECT ala.artifact_id, ala.updated_at, la.sha256
    FROM active_license_artifact ala
    LEFT JOIN license_artifacts la ON la.artifact_id = ala.artifact_id
    WHERE ala.slot = 'default'
    LIMIT 1
    `
  );
  return rows[0] ?? null;
}

async function getActiveArtifactRow(artifactId: string): Promise<LicenseArtifactRow | null> {
  const rows = await runLicenseQuery<LicenseArtifactRow>(
    `
    SELECT artifact_id,
           raw_license_text,
           sha256,
           uploaded_by_oid,
           uploaded_by_upn,
           uploaded_by_name,
           uploaded_at,
           verification_status,
           verification_error
    FROM license_artifacts
    WHERE artifact_id = $1
    LIMIT 1
    `,
    [artifactId]
  );
  return rows[0] ?? null;
}

export function clearLicenseCache() {
  cachedSummary = null;
}

export function setLicenseSummaryForTests(override: SummaryOverride | null) {
  summaryOverride = override;
  clearLicenseCache();
}

export function setLicenseQueryForTests(queryFn: LicenseQueryFn | null) {
  licenseQueryOverride = queryFn;
  clearLicenseCache();
}

export function setLicensePublicKeyForTests(value: string | null) {
  publicKeyOverride = value;
  cachedPublicKey = null;
  clearLicenseCache();
}

export async function getCurrentLicenseSummary(): Promise<EffectiveLicenseSummary> {
  if (summaryOverride) {
    return typeof summaryOverride === "function" ? summaryOverride() : summaryOverride;
  }

  const localTestingState = isLocalDockerDeployment() ? await getLocalTestingState() : null;
  const meta = localTestingState ? null : await getActiveLicenseMeta();
  const cacheKey = localTestingState
    ? `local:${localTestingState.emulateLicenseEnabled ? "enabled" : "disabled"}:${localTestingState.updatedAt || "none"}`
    : `${meta?.artifact_id || "missing"}:${meta?.sha256 || "none"}:${normalizeTimestamp(meta?.updated_at) || "none"}`;
  const ttlMs = getLicenseCacheTtlMs();

  if (cachedSummary && cachedSummary.cacheKey === cacheKey && cachedSummary.expiresAtMs > Date.now()) {
    return cachedSummary.summary;
  }

  let summary: EffectiveLicenseSummary;

  if (localTestingState) {
    summary = localTestingState.emulateLicenseEnabled
      ? getLocalDockerLicenseSummary()
      : getReadOnlyFallbackSummary("license_missing", { status: "missing", verificationStatus: "missing" });
  } else if (!meta?.artifact_id) {
    summary = getReadOnlyFallbackSummary("license_missing", { status: "missing", verificationStatus: "missing" });
  } else {
    const row = await getActiveArtifactRow(meta.artifact_id);
    if (!row) {
      summary = getReadOnlyFallbackSummary("active_license_artifact_missing_row", {
        artifactId: meta.artifact_id,
        sha256: meta.sha256,
      });
    } else {
      summary = await summarizeArtifactRow(row);
    }
  }

  cachedSummary = {
    cacheKey,
    expiresAtMs: Date.now() + ttlMs,
    summary,
  };

  return summary;
}

export async function isLicenseFeatureEnabled(featureKey: LicenseFeatureKey): Promise<boolean> {
  const summary = await getCurrentLicenseSummary();
  return Boolean(summary.features[featureKey]);
}

export async function requireLicenseFeature(featureKey: LicenseFeatureKey): Promise<EffectiveLicenseSummary> {
  const summary = await getCurrentLicenseSummary();
  if (summary.features[featureKey]) {
    return summary;
  }
  throw new LicenseFeatureError(featureKey, summary);
}

export async function insertLicenseArtifact(params: {
  rawLicenseText: string;
  actor: UploadedArtifactActor | null;
}): Promise<{
  artifactId: string;
  inspection: LicenseInspection;
}> {
  const inspection = await inspectLicenseArtifact(params.rawLicenseText);
  const actor = params.actor;
  let artifactId = "";

  await runLicenseTransaction(async (client: PoolClient) => {
    const inserted = await client.query<{ artifact_id: string }>(
      `
      INSERT INTO license_artifacts (
        raw_license_text,
        sha256,
        uploaded_by_oid,
        uploaded_by_upn,
        uploaded_by_name,
        verification_status,
        verification_error
      )
      VALUES ($1, $2, $3, $4, $5, $6, $7)
      RETURNING artifact_id
      `,
      [
        params.rawLicenseText,
        inspection.sha256,
        actor?.oid || null,
        actor?.upn || null,
        actor?.name || null,
        inspection.verificationStatus,
        inspection.verificationError,
      ]
    );
    artifactId = inserted.rows[0]?.artifact_id || "";
  });

  clearLicenseCache();

  return { artifactId, inspection };
}

export async function activateLicenseArtifact(artifactId: string): Promise<void> {
  await runLicenseTransaction(async (client: PoolClient) => {
    await client.query(
      `
      INSERT INTO active_license_artifact (slot, artifact_id, updated_at)
      VALUES ('default', $1, now())
      ON CONFLICT (slot)
      DO UPDATE SET artifact_id = EXCLUDED.artifact_id, updated_at = now()
      `,
      [artifactId]
    );
  });

  clearLicenseCache();
}

export async function clearActiveLicenseArtifact(): Promise<void> {
  await runLicenseTransaction(async (client: PoolClient) => {
    await client.query(
      `
      DELETE FROM active_license_artifact
      WHERE slot = 'default'
      `
    );
  });

  clearLicenseCache();
}
